rust/rbac-registration/src/cardano/cip509/rbac/mod.rs

```diff
@@ -16,24 +16,43 @@ use role_data::RoleData;
 use strum_macros::FromRepr;
 
 use super::types::cert_key_hash::CertKeyHash;
-use crate::utils::decode_helper::{
-    decode_any, decode_array_len, decode_bytes, decode_helper, decode_map_len,
+use crate::{
+    cardano::cip509::utils::Cip0134UriSet,
+    utils::decode_helper::{
+        decode_any, decode_array_len, decode_bytes, decode_helper, decode_map_len,
+    },
 };
 
 /// Cip509 RBAC metadata.
-#[derive(Debug, PartialEq, Clone, Default)]
+///
+/// See [this document] for more details.
+///
+/// [this document]: https://github.com/input-output-hk/catalyst-CIPs/tree/x509-role-registration-metadata/CIP-XXXX
+#[derive(Debug, PartialEq, Clone)]
 pub struct Cip509RbacMetadata {
-    /// Optional list of x509 certificates.
-    pub x509_certs: Option<Vec<X509DerCert>>,
-    /// Optional list of c509 certificates.
-    /// The value can be either the c509 certificate or c509 metadatum reference.
-    pub c509_certs: Option<Vec<C509Cert>>,
-    /// Optional list of Public keys.
-    pub pub_keys: Option<Vec<SimplePublicKeyType>>,
-    /// Optional list of revocation list.
-    pub revocation_list: Option<Vec<CertKeyHash>>,
-    /// Optional list of role data.
-    pub role_set: Option<Vec<RoleData>>,
+    /// A potentially empty list of x509 certificates.
+    pub x509_certs: Vec<X509DerCert>,
+    /// A potentially empty list of c509 certificates.
+    pub c509_certs: Vec<C509Cert>,
+    /// A set of URIs contained in both x509 and c509 certificates.
+    ///
+    /// URIs from different certificate types are stored separately and certificate
+    /// indexes are preserved too.
+    ///
+    /// This field isn't present in the encoded format and is populated by processing both
+    /// `x509_certs` and `c509_certs` fields.
+    pub certificate_uris: Cip0134UriSet,
+    /// A list of public keys that can be used instead of storing full certificates.
+    ///
+    /// Check [this section] to understand the how certificates and the public keys list
+    /// are related.
+    ///
+    /// [this section]: https://github.com/input-output-hk/catalyst-CIPs/tree/x509-role-registration-metadata/CIP-XXXX#storing-certificates-and-public-key
+    pub pub_keys: Vec<SimplePublicKeyType>,
+    /// A potentially empty list of revoked certificates.
+    pub revocation_list: Vec<CertKeyHash>,
+    /// A potentially empty list of role data.
+    pub role_set: Vec<RoleData>,
```

Comment:
Whiles its a Vec in the CBOR, there is no need to 100% preserve that format in the structure.  It should be decoded into a map.  The Key should be the Role Number, and the Value the RoleData.  We should not use `u8` as an implcit type for the `RoleNumber` we should make a `RoleNumber` type to contain it.

We should then make a set of constants of type `RoleNumber` which define our known Role Numbers, Role0 being one of those constants, rather than always using `0` as an implicit value.